README
NAME

    RotaryEncoder.h/cpp - read/decode/count rotary encoder pulses

SYNOPSIS

        #include <Arduino.h>
        #include <RotaryEncoder.h>      // https://github.com/jozef/Arduino-RotaryEncoder
    
        const uint8_t pin_a = 2; const uint8_t pin_b = 4; const uint8_t pin_btn = 7;
    
        RotaryEncoder rote(pin_a,pin_b,pin_btn);
    
        void setup () {
            rote.setup();
            Serial.begin(9600);
        }
    
        void loop () {
            if (rote.btn_pressed()) {
                // switch is pressed
                Serial.println("switch pressed");
                rote.counter = 0;
            }
            if (rote.changed()) {
                Serial.println(rote.counter);
                if (rote.counter > 0) {
                    // more steps CW then CCW
                }
            }
            delay(200);
        }

DESCRIPTION

    Arduino class/library to read-out one or more rotary encoders.

    See examples/ folder for schematics and example usage.

 PROPERTIES

    int counter:

    number of rotation steps

    int8_t direction:

    -1 for counter clockwise, 0 for one step reversed to the previous
    direction or 1 for clockwise turning.

    int limit_min; int limit_max:

    minimum and maximum value that counter can have. if both are the same
    value, no limit.

    int step:

    how much is a single rotational step. by default 1.

    uint8_t pin_a:

    Arduino pin on which encoder pin A is connected. If the pin supports
    interrupt, update() function is called automatically.

    uint8_t pin_b:

    Arduino pin on which encoder pin B is connected.

    uint8_t pin_btn:

    Arduino pin on which encoder push button is connected.

 METHODS

    RotaryEncoder(uint8_t a, uint8_t b, uint8_t btn = 0, int init_counter =
    0):

    Constructor, a and b pins are required. Button pin and initial value
    for counter optional.

    void setup():

    Will initiate pins and interrupts.

    bool update():

    Needs to be called as "fast as possible" if Arduino pin connected to
    encoder pin A is not on an interrupt pin to update state and do the
    counting.

    bool changed():

    Set to true if there was any change to the counter since the last call
    of this function.

    bool btn_pressed():

    Returns true/false if button is pressed.

EXAMPLES

 examples/RotaryEncoder_single/RotaryEncoder_synopsis.ino

    Code from SYNOPSIS section.

 examples/RotaryEncoder_single/RotaryEncoder_single.ino

    Print rotary counted value on serial port and light up led.
    Brighten/Dimm led based on rotary counter.

INSTALL

        git clone https://github.com/jozef/Arduino-RotaryEncoder.git sketchbook/libraries/RotaryEncoder

SEE ALSO

    https://blog.kutej.net/2019/01/rotary-encoder
    https://hackaday.com/2015/12/09/embed-with-elliot-debounce
    http://www.labbookpages.co.uk/electronics/debounce.html

LICENSE

    This is free software, licensed under the MIT License.

AUTHOR

        Jozef Kutej



Changelog
0.01 2019-01-08
    - initial version


RotaryEncoder.cpp
/* see README or RotaryEncoder.h for documentation */

#include <RotaryEncoder.h>
#include <Arduino.h>
#include <stdint.h>

RotaryEncoder* _first_roten = nullptr;

RotaryEncoder::RotaryEncoder(uint8_t a, uint8_t b, uint8_t btn, int init_counter)
    : pin_a(a), pin_b(b), pin_btn(btn), counter(init_counter) {
    direction = 0;
    limit_min = 0;
    limit_max = 0;
    step = 1;
    _last_counter = counter;
    _next_roten = nullptr;
}

RotaryEncoder::~RotaryEncoder() {
    RotaryEncoder** cur_roten_ptr = &_first_roten;
    while (*cur_roten_ptr != nullptr) {
        if (*cur_roten_ptr == this) {
            *cur_roten_ptr = (*cur_roten_ptr)->_next_roten;
            break;
        }
    }
}

void _RotaryEncoder_int_do_update() {
    RotaryEncoder* cur_roten = _first_roten;
    while (cur_roten != nullptr) {
        cur_roten->update();
        cur_roten = cur_roten->_next_roten;
    }
}

void RotaryEncoder::setup() {
    _next_roten = _first_roten;
    _first_roten = this;
    pinMode(pin_a, INPUT_PULLUP);
    pinMode(pin_b, INPUT_PULLUP);
    if (pin_btn) {
        pinMode(pin_btn, INPUT_PULLUP);
    }
    _last_a = digitalRead(pin_a);

    if (digitalPinToInterrupt(pin_btn) != NOT_AN_INTERRUPT) {
        attachInterrupt(digitalPinToInterrupt(pin_btn), _RotaryEncoder_int_do_update, CHANGE);
    }
    if (digitalPinToInterrupt(pin_a) != NOT_AN_INTERRUPT) {
        attachInterrupt(digitalPinToInterrupt(pin_a), _RotaryEncoder_int_do_update, CHANGE);
    }
}

bool RotaryEncoder::changed() {
    if (_last_counter != counter) {
        _last_counter = counter;
        return true;
    }
    return false;
}

bool RotaryEncoder::update() {
    uint8_t cur_a = digitalRead(pin_a);
    bool changed = false;
    if (_last_a != cur_a) {
        if (!cur_a) {
            if (digitalRead(pin_b) != cur_a) {
                if (direction <= 0) {
                    direction++;
                }
                counter += step;
            }
            else {
                if (direction >= 0) {
                    direction--;
                }
                counter -= step;
            }
            if (limit_min != limit_max) {
                if (counter < limit_min) {
                    counter = limit_min;
                }
                else if (counter > limit_max) {
                    counter = limit_max;
                }
            }
            changed = true;
        }
        _last_a = cur_a;
    }
    return changed;
}

bool RotaryEncoder::btn_pressed() {
    if (!pin_btn) {
        return false;
    }
    return !digitalRead(pin_btn);
}


RotaryEncoder.h
/* VERSION 0.01; 2018-01-08; see below for description and documentation */
#ifndef RotaryEncoder_h
#define RotaryEncoder_h

#include <stdint.h>

class RotaryEncoder {
    public:
        volatile int counter;
        volatile int8_t direction;
        int limit_min;
        int limit_max;
        int step;
        uint8_t pin_a;
        uint8_t pin_b;
        uint8_t pin_btn;
        RotaryEncoder* _next_roten;

        RotaryEncoder(uint8_t a, uint8_t b, uint8_t btn = 0, int init_counter = 0);
        ~RotaryEncoder();
        void setup();
        bool update();
        bool changed();
        bool btn_pressed();
    private:
        volatile uint8_t _last_a;
        int _last_counter;
        void _int_do_update();
};

#endif

/*

=head1 NAME

RotaryEncoder.h/cpp - read/decode/count rotary encoder pulses

=head1 SYNOPSIS

    #include <Arduino.h>
    #include <RotaryEncoder.h>      // https://github.com/jozef/Arduino-RotaryEncoder

    const uint8_t pin_a = 2; const uint8_t pin_b = 4; const uint8_t pin_btn = 7;

    RotaryEncoder rote(pin_a,pin_b,pin_btn);

    void setup () {
        rote.setup();
        Serial.begin(9600);
    }

    void loop () {
        if (rote.btn_pressed()) {
            // switch is pressed
            Serial.println("switch pressed");
            rote.counter = 0;
        }
        if (rote.changed()) {
            Serial.println(rote.counter);
            if (rote.counter > 0) {
                // more steps CW then CCW
            }
        }
        delay(200);
    }

=head1 DESCRIPTION

Arduino class/library to read-out one or more rotary encoders.

See F<examples/> folder for schematics and example usage.

=head2 PROPERTIES

int counter:

number of rotation steps

int8_t direction:

-1 for counter clockwise, 0 for one step reversed to the previous direction or 1 for clockwise turning.

int limit_min; int limit_max:

minimum and maximum value that counter can have. if both are the same value, no limit.

int step:

how much is a single rotational step. by default 1.

uint8_t pin_a:

Arduino pin on which encoder pin A is connected. If the pin supports
interrupt, update() function is called automatically.

uint8_t pin_b:

Arduino pin on which encoder pin B is connected.

uint8_t pin_btn:

Arduino pin on which encoder push button is connected.

=head2 METHODS

RotaryEncoder(uint8_t a, uint8_t b, uint8_t btn = 0, int init_counter = 0):

Constructor, a and b pins are required. Button pin and initial value
for counter optional.

void setup():

Will initiate pins and interrupts.

bool update():

Needs to be called as "fast as possible" if Arduino pin connected to encoder pin A is not on an interrupt pin
to update state and do the counting.

bool changed():

Set to true if there was any change to the counter since the last call
of this function.

bool btn_pressed():

Returns true/false if button is pressed.

=head1 EXAMPLES

=head2 examples/RotaryEncoder_single/RotaryEncoder_synopsis.ino

Code from L<SYNOPSIS> section.

=head2 examples/RotaryEncoder_single/RotaryEncoder_single.ino

Print rotary counted value on serial port and light up led.
Brighten/Dimm led based on rotary counter.

=head1 INSTALL

    git clone https://github.com/jozef/Arduino-RotaryEncoder.git sketchbook/libraries/RotaryEncoder

=head1 SEE ALSO

L<https://blog.kutej.net/2019/01/rotary-encoder>
L<https://hackaday.com/2015/12/09/embed-with-elliot-debounce>
L<http://www.labbookpages.co.uk/electronics/debounce.html>

=head1 LICENSE

This is free software, licensed under the MIT License.

=head1 AUTHOR

    Jozef Kutej

=cut

*/
